osrparse/dump.py

```diff
@@ -0,0 +1,102 @@
+import lzma
+import struct
+
+from osrparse.utils import (ReplayEventOsu, ReplayEventTaiko, ReplayEventCatch,
+    ReplayEventMania)
+from osrparse.replay import Replay
+
+def pack_byte(data: int):
+    return struct.pack("<B", data)
+
+def pack_short(data: int):
+    return struct.pack("<H", data)
+
+def pack_int(data: int):
+    return struct.pack("<I", data)
+
+def pack_long(data: int):
+    return struct.pack("<Q", data)
+
+def pack_ULEB128(data):
+    # taken from https://github.com/mohanson/leb128
+    r, i = [], len(data)
+
+    while True:
+        byte = i & 0x7f
+        i = i >> 7
+
+        if (i == 0 and byte & 0x40 == 0) or (i == -1 and byte & 0x40 != 0):
+            r.append(byte)
+            return b"".join(map(pack_byte, r))
+
+        r.append(0x80 | byte)
+
+def pack_string(data: str):
+    if data:
+        return pack_byte(11) + pack_ULEB128(data) + data.encode("utf-8")
+    return pack_byte(11) + pack_byte(0)
+
+def dump_timestamp(replay):
+    # due to converting ticks into an int, there appears to be some
+    # precision loss in the ticks value even though it's small (around 4
+    # ticks), it might be an issue with reproccessing the same replay for
+    # many times in a row
+
+    # when it's negative, it represents January 1st 0001, 01:00:00 AM UTC
+    shift = 62135596800 + 3600
+    ticks = (replay.timestamp.timestamp() + shift) * 10_000_000
+
+    return pack_long(int(ticks))
+
+
+def dump_replay_data(replay):
+    replay_data = ""
+    for event in replay.play_data:
+        if isinstance(event, ReplayEventOsu):
+            replay_data += f"{event.time_delta}|{event.x}|{event.y}|{event.keys.value},"
+        elif isinstance(event, ReplayEventTaiko):
+            replay_data += f"{event.time_delta}|{event.x}|0|{event.keys.value},"
+        elif isinstance(event, ReplayEventCatch):
+            replay_data += f"{event.time_delta}|{event.x}|0|{int(event.dashing)},"
+        elif isinstance(event, ReplayEventMania):
+            replay_data += f"{event.time_delta}|{event.keys}|0|0,"
+
+    filters = [{"id": lzma.FILTER_LZMA1, "dict_size": 2097152, "mode": lzma.MODE_FAST}]
```

Comment:
we probably want to expose `dict_size` and `mode` to consumers. In particular, consumers may want to use `lzma.MODE_NORMAL` to reduce file size - right now parsing then dumping a replay file results in a 10% *increase* in file size because we use fact compression. 

We can still keep these as the defaults (though I admit I'm not familiar with lzma decompression and don't know if 2097152 is a good dict size).

Reply:
i've used the same values used in `LegacyScoreEncoder.cs` in osu!lazerr

Reply:
I see the dict size being specified [here](https://github.com/ppy/osu/blob/master/osu.Game/Scoring/Legacy/LegacyScoreEncoder.cs#L79) (so that's probably good to leave as is, though it should be written `1 << 21` instead), but where in the code is it specifying `FILTER_LZMA1` and/or `MODE_FAST`? I can't see anywhere it specifies a mode at all, but maybe I've just missed it.

Reply:
i think i've dug into the `SharpCompress` module itself to find all the values, i don't remember tbh, sorry
`FILTER_LZMA1` is for ensuring that there will be the first and right version of LZMA used for it